docs/source/general/data.rst

```diff
@@ -2,11 +2,429 @@
+
+.. note::
+
+    ``[WIP]`` We are currently working on a new feature to make :class:`~flash.data.process.Preprocess`
+
+    and :class:`~flash.data.process.Postprocess` automatically deployable from checkpoints as
+
+    ``Endpoints`` or ``BatchTransformJob``. Stay tuned !
+
+*************************************
+How to customize existing datamodules
+*************************************
+
+Currently, Flash Tasks are implementing using :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess`.
+However, it is not a hard requirement and one can still use :class:`~torch.data.utils.Dataset`, but we highly recommend
+using :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess` instead.
+
+Example::
+
+    from flash.data.data_module import DataModule
+
+    dm = DataModule(train_dataset=MyDataset(train=True))
+    trainer = Trainer(fast_dev_run=True)
+    trainer.fit(model, data_module=dm)
+
+In order to customize Flash to your need, you need to know what are :class:`~flash.data.data_module.DataModule`
+and :class:`~flash.data.process.Preprocess` responsibilities.
+
+.. note::
+
+    At this point, we strongly encourage the readers to quickly check the :class:`~flash.data.process.Preprocess` API before getting further.
+
+The :class:`~flash.data.data_module.DataModule` provides ``classmethod`` helpers to build
+:class:`~flash.data.process.Preprocess` and :class:`~flash.data.data_pipeline.DataPipeline`,
+generate Flash Internal :class:`~flash.data.auto_dataset.AutoDataset` and populate DataLoaders with them.
+
+The :class:`~flash.data.process.Preprocess` contains the processing logic related to a given task. Users can easily override hooks
+to customize a built-in :class:`~flash.data.process.Preprocess` for their needs.
+
+Example::
+
+    from flash.vision import ImageClassificationData, ImageClassifier, ImageClassificationPreprocess
+
+    class CustomImageClassificationPreprocess(ImageClassificationPreprocess):
+
+        # Assuming you have images in numpy format,
+        # just override ``load_sample`` hook and add your own logic.
+        @staticmethod
+        def load_sample(sample) -> Tuple[Image.Image, int]:
+            # By default, ``ImageClassificationPreprocess`` expects
+            # ``.png`` or ``.jpg`` to be loaded into PIL Image.
+            numpy_image_path, label = sample
+            return np.load(numpy_image_path), sample
+
+    datamodule = ImageClassificationData.from_folders(
+        train_folder="data/hymenoptera_data/train/",
+        val_folder="data/hymenoptera_data/val/",
+        test_folder="data/hymenoptera_data/test/",
+        preprocess_cls=CustomImageClassificationPreprocess
+    )
+
+
+******************************
+Custom Preprocess + Datamodule
+******************************
+
+The example below shows a very simple ``ImageClassificationPreprocess`` with a ``ImageClassificationDataModule``.
```

Comment:
Always start with the motivation. Start by explaining it in general, and then move to the example.

If you want to add custom bla and custom bla, you might need to create custom dm and a custom datapipeline.

1. Create a DataModule - The DataModule should have helper functions to instantiate the DataModule for a relevant input type (folder structure, csv, numpy array, etc). Let's say in our example we have folders of images arranged like.... then we will create a from_folders.... etc
2. Create the preprocess- your task might need some preprocess and post process transforms. In this example.... bla bla bla
3. Make sure the processing API is used in your datamodule
4. Use your new datamodule in your task